src/inventory/beacon_inventory.rs

```diff
@@ -0,0 +1,516 @@
+            },
+            dirty: true,
+        }
+    }
+}
+
+impl Inventory for BeaconInventory {
+    fn size(&self) -> u16 {
+        self.slots.size()
+    }
+
+    fn handle_property_packet(&mut self, property: i16, value: i16) {
+        match property {
+            0 => self.info.power_level = value as u8,
+            1 => self.info.effect1 = Effect::try_from(value).ok(),
+            2 => self.info.effect2 = Effect::try_from(value).ok(),
+            _ => warn!("the server sent invalid data for the beacon"),
+        }
+        self.dirty = true;
+    }
+
+    fn id(&self) -> i32 {
+        self.id
+    }
+
+    fn get_client_state_id(&self) -> i16 {
+        self.client_state_id
+    }
+
+    fn set_client_state_id(&mut self, client_state_id: i16) {
+        self.client_state_id = client_state_id;
+        self.dirty = true;
+    }
+
+    fn get_item(&self, slot_id: u16) -> Option<Item> {
+        self.slots.get_item(slot_id)
+    }
+
+    fn set_item(&mut self, slot_id: u16, item: Option<Item>) {
+        self.slots.set_item(slot_id, item);
+        self.dirty = true;
+    }
+
+    fn get_slot(&self, x: f64, y: f64) -> Option<u16> {
+        self.slots.get_slot(x, y)
+    }
+
+    fn init(
+        &mut self,
+        renderer: Arc<Renderer>,
+        ui_container: &mut Container,
+        inventory_window: &mut InventoryWindow,
+    ) {
+        inventory_window.elements.push(vec![]); // Window textures
+        inventory_window.elements.push(vec![]); // Item slot
+        inventory_window.elements.push(vec![]); // Base slots
+        inventory_window.text_elements.push(vec![]);
+
+        let basic_elements = inventory_window.elements.get_mut(0).unwrap();
+        let icon_scale = Hud::icon_scale(renderer.clone());
+
+        let top_left_x =
+            renderer.screen_data.read().center().0 as f64 - icon_scale * WINDOW_WIDTH as f64 / 2.0;
+        let top_left_y =
+            renderer.screen_data.read().center().1 as f64 - icon_scale * WINDOW_HEIGHT as f64 / 2.0;
+
+        basic_elements.push(
+            ui::ImageBuilder::new()
+                .texture_coords((0.0, 0.0, WINDOW_WIDTH as f64, WINDOW_HEIGHT as f64))
+                .position(top_left_x, top_left_y)
+                .alignment(VAttach::Top, HAttach::Left)
+                .size(
+                    icon_scale * WINDOW_WIDTH as f64,
+                    icon_scale * WINDOW_HEIGHT as f64,
+                )
+                .texture("minecraft:gui/container/beacon")
+                .create(ui_container),
+        );
+
+        let mut pos: [(f64, f64, &str); 8] = [
+            (53.0, 22.0, "minecraft:mob_effect/speed"),
+            (77.0, 22.0, "minecraft:mob_effect/haste"),
+            (53.0, 47.0, "minecraft:mob_effect/resistance"),
+            (77.0, 47.0, "minecraft:mob_effect/jump_boost"),
+            (65.0, 72.0, "minecraft:mob_effect/strength"),
+            (144.0, 47.0, "minecraft:mob_effect/regeneration"),
+            (168.0, 47.0, ""),
```

Comment:
why is th8s empty?

Reply:
This struct helps creating the all the buttons in the ui. The first 6 buttons are the ones that always contain images of the same potion effect so i defined their images here. The other button can change its image based on the potion effect that is selected in the beacon so we need to wait for the server to send us the data of which potion effect to display first.